### PR TITLE
Forward declare Formatter

### PR DESCRIPTION
**Pull request by frutiemax**
When modifying Formatter.h, it took too long to rebuild the code. With proper forward declaring, this should be faster.

### CODE REVIEW

File: src/openrct2/core/ChecksumStream.cpp [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 #include "Endianness.h"
 
+#include <array>
```
Comment (Broxzier) [inline src/openrct2/core/ChecksumStream.cpp:14]: This is unnecessary. std::array is already included in the header, and here it's only used for a function definition that comes from the same header.

File: src/openrct2/object/WaterObject.cpp [hunk 2]
```diff
@@ -53,6 +57,7 @@ void WaterObject::Unload()
 void WaterObject::DrawPreview(rct_drawpixelinfo* dpi, int32_t width, int32_t height) const
 {
     // Write (no image)
+
```
Comment (tupaschoal) [inline src/openrct2/object/WaterObject.cpp:60]: Unrelated change, please revert.

**Review by tupaschoal — APPROVED** [review]
Looks alright to me, I'm just not very happy with the duplication of the three `Draw` functions, but I guess it's either that or some opaque pointer implementation

File: src/openrct2/drawing/Text.cpp [hunk 3]
```diff
@@ -123,12 +124,26 @@ static void DrawText(
     DrawText(dpi, coords, paint, buffer);
 }
 
+void DrawTextBasic(rct_drawpixelinfo* dpi, const ScreenCoordsXY& coords, rct_string_id format)
+{
+    Formatter ft;
```
Comment (Gymnasiast) [inline src/openrct2/drawing/Text.cpp:129]: I think this needs initialisation. (Also applies to the other two functions)

```suggestion
    Formatter ft{};
```

**Comment by Gymnasiast** [comment]
Why is this duplication/overload in the functions necessary?

**Comment by tupaschoal** [comment]
> Why is this duplication/overload in the functions necessary?

My guess is to avoid the default parameter of `Formatter = {}` which might have required the full definition to be included.

File: src/openrct2/drawing/Text.cpp (hunk 3 above)
Comment (frutiemax) [reply src/openrct2/drawing/Text.cpp]: > Why is this duplication/overload in the functions necessary?

You cannot use the constructor in the header when it is forward declared. Some function calls didn't used the same number of parameters so the duplication is needed.

**Comment by Gymnasiast** [comment]
I am reasonably satisfied with this, but I’d like the opinion of @IntelOrca , @duncanspumpkin or @ZehMatt on whether this is a good change.

Obviously, you’d also have to fix compilation for macOS and Android.

**Review by ZehMatt — APPROVED** [review]
It's actually a good change. As of the new overloads I think its fine, personally I dislike default parameter values especially when used on complex types so I'm in favor for that, there are too many stories where passing default values leads to users using the functions wrong.

**Comment by frutiemax** [comment]
I am not sure how to fix this error for the macos build:
```
/Users/runner/work/OpenRCT2/OpenRCT2/src/openrct2/platform/Shared.cpp:241:33: error: no viable constructor or deduction guide for deduction of template arguments of 'array'
    static constexpr std::array prohibited = { '/' };
```

**Comment by tupaschoal** [comment]
> I am not sure how to fix this error for the macos build:
> 
> ```
> /Users/runner/work/OpenRCT2/OpenRCT2/src/openrct2/platform/Shared.cpp:241:33: error: no viable constructor or deduction guide for deduction of template arguments of 'array'
>     static constexpr std::array prohibited = { '/' };
> ```

Try doing:
```c++
static constexpr std::array<std::string> prohibited = { '/' };
```

Or `char`
